src/main/java/fi/aalto/cs/apluscourses/intellij/actions/SubmitExerciseAction.java

```diff
@@ -0,0 +1,186 @@
+        && authenticationViewModel != null && courseViewModel != null
+        && courseViewModel.getModel() != null);
+  }
+
+  @Override
+  public void actionPerformed(@NotNull AnActionEvent e) {
+    MainViewModel mainViewModel = mainViewModelProvider.getMainViewModel(e.getProject());
+    ExerciseViewModel selectedExercise
+        = mainViewModel.exercisesViewModel.get().getSelectedExercise();
+    if (selectedExercise == null) {
+      return;
+    }
+
+    APlusAuthentication authentication
+        = mainViewModel.authenticationViewModel.get().getAuthentication();
+    Course course = mainViewModel.courseViewModel.get().getModel();
+    Project project = e.getProject();
+
+    SubmittableExercise exercise = tryGetSubmittableExercise(
+        selectedExercise.getModel(), authentication, project);
+    if (exercise == null) {
+      return;
+    }
+
+    SubmissionHistory submissionHistory = tryGetSubmissionHistory(
+        exercise, authentication, project);
+    if (submissionHistory == null) {
+      return;
+    }
+
+    List<Group> groups = tryGetUserGroups(course, authentication, project);
+    if (groups == null) {
+      return;
+    }
+
+    SubmissionViewModel viewModel = new SubmissionViewModel(
+        exercise, submissionHistory, groups, authentication, project);
+
+    if (!new ModuleSelectionDialog(viewModel).showAndGet()
+        || viewModel.getSelectedModule() == null) {
+      return;
+    }
+
+    List<Path> filePaths = tryGetFilePaths(
+        exercise.getFilenames(), viewModel.getSelectedModule(), project);
+    if (filePaths == null) {
+      return;
+    }
+
+    if (new SubmissionDialog(viewModel).showAndGet()) {
+      // Do actual submission
+    }
+  }
+
+  @Nullable
+  private SubmittableExercise tryGetSubmittableExercise(@NotNull Exercise exercise,
+                                                        @NotNull APlusAuthentication authentication,
+                                                        @Nullable Project project) {
+    try {
+      SubmittableExercise submittableExercise = SubmittableExercise.fromExerciseId(
+          exercise.getId(), authentication);
+      if (submittableExercise.getFilenames().isEmpty()) {
+        notifier.notify(new NotSubmittableNotification(), project);
+        return null;
+      }
+      return submittableExercise;
+    } catch (IOException e) {
+      notifyNetworkError(e, project);
+      return null;
+    }
+  }
+
+  @Nullable
+  private SubmissionHistory tryGetSubmissionHistory(@NotNull Exercise exercise,
+                                                    @NotNull APlusAuthentication authentication,
+                                                    @Nullable Project project) {
+    try {
+      return SubmissionHistory.getSubmissionHistory(exercise.getId(), authentication);
+    } catch (IOException e) {
+      notifyNetworkError(e, project);
+      return null;
+    }
+  }
```

Comment:
Do we want to have some caching here and in `tryGetUserGroups` (not inside the method but somewhere)?

Reply:
Probably yes, but I would start by caching the exercise requests (especially the request for all exercises), as those are much larger

Reply:
And of course caching the number of previous submissions runs the risk of telling the student that they are making submission 5/10 when they are actually making submission 10/10 :smile: